pymeasure/instruments/common_base.py

```diff
@@ -0,0 +1,569 @@
+            for attr in obj.__dict__:
+                if attr in special_names:
+                    # Copy class special variable at instance level, prefixing reserved_prefix
+                    setattr(self, self.__reserved_prefix + attr, obj.__dict__[attr])
+        return special_names
+
+    def __setattr__(self, name, value):
+        """ Add reserved_prefix in front of special variables."""
+        if hasattr(self, '_special_names'):
+            if name in self._special_names:
+                name = self.__reserved_prefix + name
+        super().__setattr__(name, value)
+
+    def __getattribute__(self, name):
+        """ Prevent read access to variables with special names used to
+        support dynamic property behaviour."""
+        if name in ('_special_names', '__dict__'):
+            return super().__getattribute__(name)
+        if hasattr(self, '_special_names'):
+            if name in self._special_names:
+                raise AttributeError(
+                    f"{name} is a reserved variable name and it cannot be read")
+        return super().__getattribute__(name)
+
+    # Channel management
+    def add_child(self, cls, id, collection="channels", prefix="ch", **kwargs):
+        """Add a child to this instance and return its index in the children list.
+
+        The newly created child may be accessed either by the index in the
+        children list or by the created attribute.
+        The fifth channel of `instrument` with id "F" has two access options:
+        :code:`instrument.channels[4] == instrument.ch_F`
+
+        .. note::
+
+            Do not change the default `collection` or `prefix` parameter, unless
+            you have to distinguish several collections of different children,
+            e.g. different channel types (analog and digital).
+
+        :param cls: Class of the channel.
+        :param id: Child id how it is used in communication, e.g. `"A"`.
+            If `None`, the child will be added directly under the collection name.
+        :param collection: Name of the collection of children, used for the list.
+        :param prefix: Collection prefix for the attributes, e.g. `"ch"`
+            creates attribute `self.ch_A`. An underscore separates prefix from id.
+        :param \\**kwargs: Keyword arguments for the channel creator.
+        :returns: Index of this instance's channels.
+        """
+        child = cls(self, id, **kwargs)
+        collection_list = getattr(self, collection, [])
+        if id is None:
+            if collection_list:
+                raise ValueError(f"An attribute '{collection}' already exists.")
+            setattr(self, collection, child)
+            child._name = collection
+            child._collection = None
+            return 0
+        else:
+            if not collection_list or isinstance(collection_list, ChildDescriptor):
+                # Add a grouplist to the parent.
+                setattr(self, collection, collection_list)
```

Comment:
I don't get the setattr for the second part of the if -- if `collection_list is a ChildDescriptor, it had already existed at `collection` from the start, why set it again to the same name?